### PR TITLE
Hide complexity based on lecture examples  Issue #244

### PR DESCRIPTION
**Pull request by maedae**

### CODE REVIEW

File: app/presenters/tutorial_view_chapter_presenter.rb [hunk 1]
```diff
@@ -0,0 +1,18 @@
+class ChapterPresenter
```
Comment (sumeetjain) [inline app/presenters/tutorial_view_chapter_presenter.rb:1]: Has no descriptive comment (IrresponsibleModule)

File: app/presenters/tutorial_view_step_presenter.rb [hunk 2]
```diff
@@ -0,0 +1,31 @@
+class StepPresenter
+    @@counter = 0
+    @@chapter = 0
```
Comment (sumeetjain) [inline app/presenters/tutorial_view_step_presenter.rb]: Declares the class variable @@chapter (ClassVariable)

File: app/presenters/tutorial_view_step_presenter.rb [hunk 3]
```diff
@@ -0,0 +1,31 @@
+class StepPresenter
+    @@counter = 0
```
Comment (sumeetjain) [inline app/presenters/tutorial_view_step_presenter.rb]: Declares the class variable @@counter (ClassVariable)

File: app/presenters/tutorial_view_step_presenter.rb [hunk 4]
```diff
@@ -0,0 +1,31 @@
...
+
+  def step
+    @step
+  end
+
+
+
+  def content
+    if self.step.warning?
+      "WARNING! #{step.title} - (#{step.video_time_conversion_for_steps})"
```
Comment (sumeetjain) [inline app/presenters/tutorial_view_step_presenter.rb]: Calls step.title 2 times (DuplicateMethodCall)

Comment (sumeetjain) [inline app/presenters/tutorial_view_step_presenter.rb]: Calls step.video_time_conversion_for_steps 2 times (DuplicateMethodCall)

File: app/presenters/tutorial_view_step_presenter.rb [hunk 5]
```diff
@@ -0,0 +1,31 @@
+class StepPresenter
```
Comment (sumeetjain) [inline app/presenters/tutorial_view_step_presenter.rb]: Has no descriptive comment (IrresponsibleModule)

File: app/presenters/tutorial_view_chapter_presenter.rb [hunk 6]
```diff
@@ -0,0 +1,18 @@
...
+  def initialize(chapter_object)
+    @chapter = chapter_object
+  end
+
+  def chapter
```
Comment (sumeetjain) [inline app/presenters/tutorial_view_chapter_presenter.rb:7]: Use `attr_reader` to define trivial reader methods.

File: app/presenters/tutorial_view_chapter_presenter.rb [hunk 7]
```diff
@@ -0,0 +1,18 @@
...
+    @chapter
+  end
+
+  def content
+      "Chapter #{chapter_number}: #{chapter.title} - (#{chapter.video_time_conversion_for_chapters})"
```
Comment (sumeetjain) [inline app/presenters/tutorial_view_chapter_presenter.rb]: Use 2 (not 4) spaces for indentation.

Comment (sumeetjain) [inline app/presenters/tutorial_view_chapter_presenter.rb]: Line is too long. [101/80]

File: app/presenters/tutorial_view_step_presenter.rb (hunk 3 above)
Comment (sumeetjain) [inline app/presenters/tutorial_view_step_presenter.rb]: Use 2 (not 4) spaces for indentation.

Comment (sumeetjain) [inline app/presenters/tutorial_view_step_presenter.rb]: Replace class var @@counter with a class instance var.

File: app/presenters/tutorial_view_step_presenter.rb (hunk 2 above)
Comment (sumeetjain) [inline app/presenters/tutorial_view_step_presenter.rb]: Replace class var @@chapter with a class instance var.

File: app/presenters/tutorial_view_step_presenter.rb [hunk 8]
```diff
@@ -0,0 +1,31 @@
...
+    @@counter = 0
+    @@chapter = 0
+  def initialize(step_object, chapter_id)
```
Comment (sumeetjain) [inline app/presenters/tutorial_view_step_presenter.rb]: Inconsistent indentation detected.

File: app/presenters/tutorial_view_step_presenter.rb [hunk 9]
```diff
@@ -0,0 +1,31 @@
...
+    @step = step_object
+    @current_chapter = chapter_id
+
+    if @@chapter != @current_chapter
+       @@counter = 0
```
Comment (sumeetjain) [inline app/presenters/tutorial_view_step_presenter.rb]: Use 2 (not 3) spaces for indentation.

Comment (sumeetjain) [inline app/presenters/tutorial_view_step_presenter.rb]: Replace class var @@counter with a class instance var.

File: app/presenters/tutorial_view_step_presenter.rb [hunk 10]
```diff
@@ -0,0 +1,31 @@
...
+       @@chapter = @current_chapter
```
Comment (sumeetjain) [inline app/presenters/tutorial_view_step_presenter.rb]: Replace class var @@chapter with a class instance var.

File: app/presenters/tutorial_view_step_presenter.rb [hunk 11]
```diff
@@ -0,0 +1,31 @@
...
+    end
+
```
Comment (sumeetjain) [inline app/presenters/tutorial_view_step_presenter.rb]: Extra empty line detected at method body end.

File: app/presenters/tutorial_view_step_presenter.rb [hunk 12]
```diff
@@ -0,0 +1,31 @@
...
+  end
+
+
```
Comment (sumeetjain) [inline app/presenters/tutorial_view_step_presenter.rb]: Extra blank line detected.

File: app/presenters/tutorial_view_step_presenter.rb [hunk 13]
```diff
@@ -0,0 +1,31 @@
...
+  def step
```
Comment (sumeetjain) [inline app/presenters/tutorial_view_step_presenter.rb]: Inconsistent indentation detected.

Comment (sumeetjain) [inline app/presenters/tutorial_view_step_presenter.rb]: Use `attr_reader` to define trivial reader methods.

File: app/presenters/tutorial_view_step_presenter.rb [hunk 14]
```diff
@@ -0,0 +1,31 @@
...
+    @step
+  end
+
+
```
Comment (sumeetjain) [inline app/presenters/tutorial_view_step_presenter.rb]: Extra blank line detected.

File: app/presenters/tutorial_view_step_presenter.rb [hunk 15]
```diff
@@ -0,0 +1,31 @@
...
+
```
Comment (sumeetjain) [inline app/presenters/tutorial_view_step_presenter.rb]: Extra blank line detected.

File: app/presenters/tutorial_view_step_presenter.rb [hunk 16]
```diff
@@ -0,0 +1,31 @@
...
+  def content
```
Comment (sumeetjain) [inline app/presenters/tutorial_view_step_presenter.rb]: Inconsistent indentation detected.

File: app/presenters/tutorial_view_step_presenter.rb [hunk 17]
```diff
@@ -0,0 +1,31 @@
...
+    if self.step.warning?
+      "WARNING! #{step.title} - (#{step.video_time_conversion_for_steps})"
+    else
+      @@counter += 1
```
Comment (sumeetjain) [inline app/presenters/tutorial_view_step_presenter.rb]: Replace class var @@counter with a class instance var.

File: app/presenters/tutorial_view_step_presenter.rb [hunk 18]
```diff
@@ -0,0 +1,31 @@
...
+      "Step #{@@counter}: #{step.title} - (#{step.video_time_conversion_for_steps})"
```
Comment (sumeetjain) [inline app/presenters/tutorial_view_step_presenter.rb]: Line is too long. [84/80]

File: app/presenters/tutorial_view_step_presenter.rb [hunk 19]
```diff
@@ -0,0 +1,31 @@
...
+    end
+  end
+
```
Comment (sumeetjain) [inline app/presenters/tutorial_view_step_presenter.rb]: Extra empty line detected at class body end.

File: app/presenters/tutorial_view_chapter_presenter.rb [hunk 20]
```diff
@@ -0,0 +1,15 @@
+class ChapterPresenter
+  def initialize(chapter_object, chapter_index)
```
Comment (sumeetjain) [inline app/presenters/tutorial_view_chapter_presenter.rb:2]: Has unused parameter 'chapter_index' (UnusedParameters)

File: app/presenters/tutorial_view_chapter_presenter.rb [hunk 21]
```diff
@@ -0,0 +1,15 @@
...
+    @chapter = chapter_object
+    @count = chapter_index += 1
```
Comment (sumeetjain) [inline app/presenters/tutorial_view_chapter_presenter.rb:4]: Useless assignment to variable - `chapter_index`.

File: app/presenters/tutorial_view_chapter_presenter.rb [hunk 22]
```diff
@@ -0,0 +1,15 @@
...
+  end
+
+  def chapter
```
Comment (sumeetjain) [inline app/presenters/tutorial_view_chapter_presenter.rb:7]: Use `attr_reader` to define trivial reader methods.

File: app/presenters/tutorial_view_chapter_presenter.rb [hunk 23]
```diff
@@ -0,0 +1,15 @@
...
+    @chapter
+  end
+
+  def content
+      "Chapter #{@count}: #{chapter.title} - (#{chapter.video_time_conversion_for_chapters})"
```
Comment (sumeetjain) [inline app/presenters/tutorial_view_chapter_presenter.rb]: Use 2 (not 4) spaces for indentation.

Comment (sumeetjain) [inline app/presenters/tutorial_view_chapter_presenter.rb]: Line is too long. [93/80]

File: app/presenters/tutorial_view_chapter_presenter.rb [hunk 24]
```diff
@@ -0,0 +1,15 @@
...
+  end
+
```
Comment (sumeetjain) [inline app/presenters/tutorial_view_chapter_presenter.rb:18]: Extra empty line detected at class body end.

**Comment by sumeetjain** [comment]
Just spitballin', but...

What if `ChapterPresenter` had some of this:

``` ruby
def initialize(...)
  # ...
  @steps = @chapter.steps.map { |s| StepPresenter.new(s) }
end
```

So now a `ChapterPresenter` object contains a collection of all of its chapter's steps–as StepPresenter Objects.

And in your view, you can loop through `@chapter_presenter.steps` and call presenter methods on each step. But the added ability now is that each step can access information from its ChapterPresenter.

I feel like the ChapterPresenter should be able to know about the number ordering of its steps/warnings, and it can pass that knowledge into each of its `steps`.

**Comment by maedae** [comment]
Yup! that was just what Wendy and I were talking about. I was considering making an array of the proper strings in chapters ["Step 1", "Step 2", "WARNING", "WARNING", Step3"] and returning the value in the chapters array based on the step's index, since both are indexes related to the same instance

File: app/presenters/tutorial_view_step_collection_presenter.rb [hunk 25]
```diff
@@ -0,0 +1,69 @@
...
+    @steps = step_object_arr
+    @counter = 0
+    @steps_info_arr = []
+    @main_steps_arr = []
+  end
+
+  def set_step(step)
+    @step = step
+  end
+
```
Comment (sumeetjain) [inline app/presenters/tutorial_view_step_collection_presenter.rb:12]: Remove trailing whitespace

File: app/views/tutorials/view.html.erb [hunk 26]
```diff
@@ -1,51 +1,61 @@
-<h1><%= @tutorial.title %></h1>
-<%= "Access: " + @tutorial.get_privacy_notice %><br>
- <%= "Category: " + @tutorial.get_category_notice %> <br>
-<%= @tutorial.description %><br><br>
+<div class="tutorial_info">
```
Comment (sumeetjain) [inline app/views/tutorials/view.html.erb:1]: Remove trailing whitespace

File: app/presenters/tutorial_view_step_collection_presenter.rb [hunk 27]
```diff
@@ -0,0 +1,69 @@
+class StepCollectionPresenter
```
Comment (sumeetjain) [inline app/presenters/tutorial_view_step_collection_presenter.rb:1]: Has no descriptive comment (IrresponsibleModule)

Comment (sumeetjain) [inline app/presenters/tutorial_view_step_collection_presenter.rb:1]: Has at least 5 instance variables (TooManyInstanceVariables)

File: app/presenters/tutorial_view_step_collection_presenter.rb [hunk 28]
```diff
@@ -0,0 +1,69 @@
...
+
+  def get_description()
+    @step.description
+  end
+  
+  def content
+      "#{@step.title} - (#{@step.video_time_conversion_for_steps})"
+  end
+
+  def get_steps_ordering_array()
```
Comment (sumeetjain) [inline app/presenters/tutorial_view_step_collection_presenter.rb:49]: Has approx 12 statements (TooManyStatements)

File: app/presenters/tutorial_view_chapter_presenter.rb [hunk 29]
```diff
@@ -0,0 +1,19 @@
...
+    @chapter = chapter_object
+    @count = chapter_index += 1
+  end
+
+  def chapter
+    @chapter
+  end
+
+  def content
+       "Chapter #{@count}: #{chapter.title} - (#{chapter.video_time_conversion_for_chapters})"
```
Comment (sumeetjain) [inline app/presenters/tutorial_view_chapter_presenter.rb:12]: Use 2 (not 5) spaces for indentation.

Comment (sumeetjain) [inline app/presenters/tutorial_view_chapter_presenter.rb:12]: Line is too long. [94/80]

File: app/presenters/tutorial_view_chapter_presenter.rb [hunk 30]
```diff
@@ -0,0 +1,19 @@
...
+  end
+
+  def get_description()
```
Comment (sumeetjain) [inline app/presenters/tutorial_view_chapter_presenter.rb:15]: Do not prefix reader method names with `get_`.

Comment (sumeetjain) [inline app/presenters/tutorial_view_chapter_presenter.rb:15]: Omit the parentheses in defs when the method doesn't accept any arguments.

File: app/presenters/tutorial_view_chapter_presenter.rb [hunk 31]
```diff
@@ -0,0 +1,19 @@
...
+   return  @chapter.description
```
Comment (sumeetjain) [inline app/presenters/tutorial_view_chapter_presenter.rb:16]: Use 2 (not 1) spaces for indentation.

Comment (sumeetjain) [inline app/presenters/tutorial_view_chapter_presenter.rb:16]: Unnecessary spacing detected.

File: app/presenters/tutorial_view_chapter_presenter.rb [hunk 32]
```diff
@@ -0,0 +1,19 @@
...
+  end
+
```
Comment (sumeetjain) [inline app/presenters/tutorial_view_chapter_presenter.rb:18]: Extra empty line detected at class body end.

File: app/presenters/tutorial_view_step_collection_presenter.rb [hunk 33]
```diff
@@ -0,0 +1,69 @@
+class StepCollectionPresenter
+  def initialize(step_object_arr)
+    @steps = step_object_arr
+    @counter = 0
+    @steps_info_arr = []
+    @main_steps_arr = []
+  end
+
+  def set_step(step)
```
Comment (sumeetjain) [inline app/presenters/tutorial_view_step_collection_presenter.rb:9]: Do not prefix writer method names with `set_`.

File: app/presenters/tutorial_view_step_collection_presenter.rb (hunk 25 above)
Comment (sumeetjain) [inline app/presenters/tutorial_view_step_collection_presenter.rb:12]: Trailing whitespace detected.

File: app/presenters/tutorial_view_step_collection_presenter.rb [hunk 34]
```diff
@@ -0,0 +1,69 @@
...
+    @step = step
+  end
+  
+  def get_step
```
Comment (sumeetjain) [inline app/presenters/tutorial_view_step_collection_presenter.rb:13]: Do not prefix reader method names with `get_`.

File: app/presenters/tutorial_view_step_collection_presenter.rb [hunk 35]
```diff
@@ -0,0 +1,69 @@
...
+    @step
+  end
+
+  def empty_check
+    if @steps.empty?
+      get_main_steps_arr()
```
Comment (sumeetjain) [inline app/presenters/tutorial_view_step_collection_presenter.rb:19]: Do not use parentheses for method calls with no arguments.

File: app/presenters/tutorial_view_step_collection_presenter.rb [hunk 36]
```diff
@@ -0,0 +1,69 @@
...
+    else
+      get_steps_ordering_array()
```
Comment (sumeetjain) [inline app/presenters/tutorial_view_step_collection_presenter.rb:21]: Do not use parentheses for method calls with no arguments.

File: app/presenters/tutorial_view_step_collection_presenter.rb [hunk 37]
```diff
@@ -0,0 +1,69 @@
...
+    end
+  end
+
+  def reset_steps_info_arr
+    @steps_info_arr = []
+  end
+
+  def get_main_steps_arr()
```
Comment (sumeetjain) [inline app/presenters/tutorial_view_step_collection_presenter.rb:29]: Do not prefix reader method names with `get_`.

Comment (sumeetjain) [inline app/presenters/tutorial_view_step_collection_presenter.rb:29]: Omit the parentheses in defs when the method doesn't accept any arguments.

File: app/presenters/tutorial_view_step_collection_presenter.rb [hunk 38]
```diff
@@ -0,0 +1,69 @@
...
+    return @main_steps_arr
+  end
+
+  def get_start_time()
```
Comment (sumeetjain) [inline app/presenters/tutorial_view_step_collection_presenter.rb:33]: Do not prefix reader method names with `get_`.

Comment (sumeetjain) [inline app/presenters/tutorial_view_step_collection_presenter.rb:33]: Omit the parentheses in defs when the method doesn't accept any arguments.

File: app/presenters/tutorial_view_step_collection_presenter.rb [hunk 39]
```diff
@@ -0,0 +1,69 @@
...
+      @step.start_time
```
Comment (sumeetjain) [inline app/presenters/tutorial_view_step_collection_presenter.rb:34]: Use 2 (not 4) spaces for indentation.

File: app/presenters/tutorial_view_step_collection_presenter.rb [hunk 40]
```diff
@@ -0,0 +1,69 @@
...
+  end
+
+  def get_end_time()
```
Comment (sumeetjain) [inline app/presenters/tutorial_view_step_collection_presenter.rb:37]: Do not prefix reader method names with `get_`.

Comment (sumeetjain) [inline app/presenters/tutorial_view_step_collection_presenter.rb:37]: Omit the parentheses in defs when the method doesn't accept any arguments.

File: app/presenters/tutorial_view_step_collection_presenter.rb [hunk 41]
```diff
@@ -0,0 +1,69 @@
...
+      @step.end_time
```
Comment (sumeetjain) [inline app/presenters/tutorial_view_step_collection_presenter.rb:38]: Use 2 (not 4) spaces for indentation.

File: app/presenters/tutorial_view_step_collection_presenter.rb [hunk 42]
```diff
@@ -0,0 +1,69 @@
...
+  end
+
+  def get_description()
```
Comment (sumeetjain) [inline app/presenters/tutorial_view_step_collection_presenter.rb:41]: Do not prefix reader method names with `get_`.

Comment (sumeetjain) [inline app/presenters/tutorial_view_step_collection_presenter.rb:41]: Omit the parentheses in defs when the method doesn't accept any arguments.

File: app/presenters/tutorial_view_step_collection_presenter.rb [hunk 43]
```diff
@@ -0,0 +1,69 @@
...
+    @step.description
+  end
+
```
Comment (sumeetjain) [inline app/presenters/tutorial_view_step_collection_presenter.rb:44]: Trailing whitespace detected.

File: app/presenters/tutorial_view_step_collection_presenter.rb [hunk 44]
```diff
@@ -0,0 +1,69 @@
...
+  def get_end_time()
+      @step.end_time
+  end
+
+  def get_description()
+    @step.description
+  end
+  
+  def content
+      "#{@step.title} - (#{@step.video_time_conversion_for_steps})"
```
Comment (sumeetjain) [inline app/presenters/tutorial_view_step_collection_presenter.rb:46]: Use 2 (not 4) spaces for indentation.

File: app/presenters/tutorial_view_step_collection_presenter.rb (hunk 28 above)
Comment (sumeetjain) [inline app/presenters/tutorial_view_step_collection_presenter.rb:49]: Assignment Branch Condition size for get_steps_ordering_array is too high. [16.16/15]

Comment (sumeetjain) [inline app/presenters/tutorial_view_step_collection_presenter.rb:49]: Method has too many lines. [16/10]

Comment (sumeetjain) [inline app/presenters/tutorial_view_step_collection_presenter.rb:49]: Do not prefix reader method names with `get_`.

Comment (sumeetjain) [inline app/presenters/tutorial_view_step_collection_presenter.rb:49]: Omit the parentheses in defs when the method doesn't accept any arguments.

File: app/presenters/tutorial_view_step_collection_presenter.rb [hunk 45]
```diff
@@ -0,0 +1,69 @@
...
+  def get_steps_ordering_array()
+    @steps.each do |step|
+      set_step(step)
+      reset_steps_info_arr
+
+      if step.warning?
+        @step_info_arr << "WARNING!"
+      else
+        @counter += 1
+        @steps_info_arr <<  "STEP: #{@counter} "
```
Comment (sumeetjain) [inline app/presenters/tutorial_view_step_collection_presenter.rb:58]: Operator `<<` should be surrounded by a single space.

Comment (sumeetjain) [inline app/presenters/tutorial_view_step_collection_presenter.rb:58]: Unnecessary spacing detected.

File: app/presenters/tutorial_view_step_collection_presenter.rb [hunk 46]
```diff
@@ -0,0 +1,69 @@
...
+      end
+        @steps_info_arr << content()
```
Comment (sumeetjain) [inline app/presenters/tutorial_view_step_collection_presenter.rb:60]: Inconsistent indentation detected.

Comment (sumeetjain) [inline app/presenters/tutorial_view_step_collection_presenter.rb:60]: Do not use parentheses for method calls with no arguments.

File: app/presenters/tutorial_view_step_collection_presenter.rb [hunk 47]
```diff
@@ -0,0 +1,69 @@
...
+        @steps_info_arr << get_start_time()
```
Comment (sumeetjain) [inline app/presenters/tutorial_view_step_collection_presenter.rb:61]: Inconsistent indentation detected.

Comment (sumeetjain) [inline app/presenters/tutorial_view_step_collection_presenter.rb:61]: Do not use parentheses for method calls with no arguments.

File: app/presenters/tutorial_view_step_collection_presenter.rb [hunk 48]
```diff
@@ -0,0 +1,69 @@
...
+        @steps_info_arr << get_end_time()
```
Comment (sumeetjain) [inline app/presenters/tutorial_view_step_collection_presenter.rb:62]: Inconsistent indentation detected.

Comment (sumeetjain) [inline app/presenters/tutorial_view_step_collection_presenter.rb:62]: Do not use parentheses for method calls with no arguments.

File: app/presenters/tutorial_view_step_collection_presenter.rb [hunk 49]
```diff
@@ -0,0 +1,69 @@
...
+        @steps_info_arr << get_description()
```
Comment (sumeetjain) [inline app/presenters/tutorial_view_step_collection_presenter.rb:63]: Inconsistent indentation detected.

Comment (sumeetjain) [inline app/presenters/tutorial_view_step_collection_presenter.rb:63]: Do not use parentheses for method calls with no arguments.

File: app/presenters/tutorial_view_step_collection_presenter.rb [hunk 50]
```diff
@@ -0,0 +1,69 @@
...
+        @main_steps_arr << @steps_info_arr
```
Comment (sumeetjain) [inline app/presenters/tutorial_view_step_collection_presenter.rb:64]: Inconsistent indentation detected.

File: app/presenters/tutorial_view_step_collection_presenter.rb [hunk 51]
```diff
@@ -0,0 +1,69 @@
...
+    end
+    get_main_steps_arr()
```
Comment (sumeetjain) [inline app/presenters/tutorial_view_step_collection_presenter.rb:66]: Do not use parentheses for method calls with no arguments.

File: app/presenters/tutorial_view_step_collection_presenter.rb [hunk 52]
```diff
@@ -0,0 +1,69 @@
...
+  end
+
```
Comment (sumeetjain) [inline app/presenters/tutorial_view_step_collection_presenter.rb:68]: Extra empty line detected at class body end.

**Pull request merged**